File: server/module/module-erp/erp-server/src/service/erp_purchase_order_attachment.rs
use std::clone;

use file_common::service::system_file;
use sea_orm::{ActiveModelTrait, ColumnTrait, Condition, DatabaseConnection, DatabaseTransaction, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder};
use crate::model::erp_purchase_order_attachment::{Model as ErpPurchaseOrderAttachmentModel, ActiveModel as ErpPurchaseOrderAttachmentActiveModel, Entity as ErpPurchaseOrderAttachmentEntity, Column};
use erp_model::request::erp_purchase_order_attachment::{CreateErpPurchaseOrderAttachmentRequest, UpdateErpPurchaseOrderAttachmentRequest, PaginatedKeywordRequest};
use erp_model::response::erp_purchase_order_attachment::{ErpPurchaseOrderAttachmentBaseResponse, ErpPurchaseOrderAttachmentResponse};
use crate::convert::erp_purchase_order_attachment::{create_request_to_model, model_to_base_response, model_to_response, update_request_to_model};
use anyhow::{anyhow, Context, Result};
use sea_orm::ActiveValue::Set;
use common::constants::enum_constants::{STATUS_DISABLE, STATUS_ENABLE};
use common::base::page::PaginatedResponse;
use common::context::context::LoginUserContext;
use common::interceptor::orm::active_filter::ActiveFilterEntityTrait;

pub async fn create(db: &DatabaseConnection, login_user: LoginUserContext, request: CreateErpPurchaseOrderAttachmentRequest) -> Result<i64> {
    let mut erp_purchase_order_attachment = create_request_to_model(&request);
    erp_purchase_order_attachment.creator = Set(Some(login_user.id));
    erp_purchase_order_attachment.updater = Set(Some(login_user.id));
    erp_purchase_order_attachment.tenant_id = Set(login_user.tenant_id);
    let erp_purchase_order_attachment = erp_purchase_order_attachment.insert(db).await?;
    Ok(erp_purchase_order_attachment.id)
}

pub async fn create_batch(db: &DatabaseConnection, txn: &DatabaseTransaction, login_user: LoginUserContext, purchase_id: i64, requests: Vec<CreateErpPurchaseOrderAttachmentRequest>) -> Result<()> {
    let file_ids: Vec<i64> = requests.iter().clone().map(|request| request.file_id).collect();

    let models: Vec<ErpPurchaseOrderAttachmentActiveModel> = requests
        .into_iter()
        .map(|request| {
            let mut model = create_request_to_model(&request);
            model.purchase_id = Set(purchase_id);
            model.department_id = Set(login_user.department_id);
            model.department_code = Set(login_user.department_code.clone());
            model.creator = Set(Some(login_user.id));
            model.updater = Set(Some(login_user.id));
            model.tenant_id = Set(login_user.tenant_id);
            model
        })
        .collect();

    if !models.is_empty() {
        ErpPurchaseOrderAttachmentEntity::insert_many(models)
            .exec(txn)
            .await
            .with_context(|| "Failed to save attachment")?;
    }

    // 启用文件
    system_file::enable_outer(&db, &txn, login_user, file_ids).await?;
    Ok(())
}

pub async fn update(db: &DatabaseConnection, login_user: LoginUserContext, request: UpdateErpPurchaseOrderAttachmentRequest) -> Result<()> {
    let erp_purchase_order_attachment = ErpPurchaseOrderAttachmentEntity::find_by_id(request.id)
        .one(db)
        .await?
        .ok_or_else(|| anyhow!("记录未找到"))?;

    let mut erp_purchase_order_attachment = update_request_to_model(&request, erp_purchase_order_attachment);
    erp_purchase_order_attachment.updater = Set(Some(login_user.id));
    erp_purchase_order_attachment.update(db).await?;
    Ok(())
}

pub async fn delete(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let erp_purchase_order_attachment = ErpPurchaseOrderAttachmentActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        deleted: Set(true),
        ..Default::default()
    };
    erp_purchase_order_attachment.update(db).await?;
    Ok(())
}

pub async fn get_by_id(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<Option<ErpPurchaseOrderAttachmentResponse>> {
    let condition = Condition::all()
            .add(Column::Id.eq(id))
            .add(Column::TenantId.eq(login_user.tenant_id));
            
    let erp_purchase_order_attachment = ErpPurchaseOrderAttachmentEntity::find_active_with_condition(condition)
        .one(db).await?;
    Ok(erp_purchase_order_attachment.map(model_to_response))
}

pub async fn get_paginated(db: &DatabaseConnection, login_user: LoginUserContext, params: PaginatedKeywordRequest) -> Result<PaginatedResponse<ErpPurchaseOrderAttachmentResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let paginator = ErpPurchaseOrderAttachmentEntity::find_active_with_condition(condition)
        .order_by_desc(Column::UpdateTime)
        .paginate(db, params.base.size);

    let total = paginator.num_items().await?;
    let total_pages = (total + params.base.size - 1) / params.base.size; // 向上取整
    let list = paginator
        .fetch_page(params.base.page - 1) // SeaORM 页码从 0 开始，所以减 1
        .await?
        .into_iter()
        .map(model_to_response)
        .collect();

    Ok(PaginatedResponse {
        list,
        total_pages,
        page: params.base.page,
        size: params.base.size,
        total,
    })
}

pub async fn list(db: &DatabaseConnection, login_user: LoginUserContext) -> Result<Vec<ErpPurchaseOrderAttachmentResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let list = ErpPurchaseOrderAttachmentEntity::find_active_with_condition(condition)
        .all(db).await?;
    Ok(list.into_iter().map(model_to_response).collect())
}

pub async fn list_by_purchase_id(db: &DatabaseConnection, login_user: LoginUserContext, purchase_id: i64) -> Result<Vec<ErpPurchaseOrderAttachmentBaseResponse>> {
    let list = ErpPurchaseOrderAttachmentEntity::find_active_with_data_permission(login_user.clone())
        .filter(Column::TenantId.eq(login_user.tenant_id))
        .filter(Column::PurchaseId.eq(purchase_id))
        .all(db).await?;
    Ok(list.into_iter().map(model_to_base_response).collect())
}

File: admin/src/api/erp_purchase_order_attachment.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { UploadFile } from '@/components/CustomizedFileUpload';
import { api } from '@/utils/request';

const apis = {
  create: '/erp/erp_purchase_order_attachment/create', // 新增
  update: '/erp/erp_purchase_order_attachment/update', // 修改
  delete: '/erp/erp_purchase_order_attachment/delete', // 删除
  get: '/erp/erp_purchase_order_attachment/get', // 单条查询
  list: '/erp/erp_purchase_order_attachment/list', // 列表查询
  page: '/erp/erp_purchase_order_attachment/page', // 分页查询
  list_purchase: '/erp/erp_purchase_order_attachment/list_purchase', // 采购订单附件列表
}

export interface ErpPurchaseOrderAttachmentRequest {
  id?: number; // 附件ID
  purchase_id?: number; // 采购订单ID
  file_id?: number; // 文件ID
  remarks?: string; // 备注

  file?: UploadFile | null;
}

export interface ErpPurchaseOrderAttachmentResponse {
  id: number; // 附件ID
  purchase_id: number; // 采购订单ID
  file_id: number; // 文件ID
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间
}

export interface ErpPurchaseOrderAttachmentBaseResponse {
  id: number; // 附件ID
  purchase_id: number; // 采购订单ID
  file_id: number; // 文件ID
  file_name: string; // 文件名
  remarks?: string; // 备注
}

export interface ErpPurchaseOrderAttachmentQueryCondition extends PaginatedRequest {

}

export const createErpPurchaseOrderAttachment = (erp_purchase_order_attachment: ErpPurchaseOrderAttachmentRequest): Promise<number> => {
  return api.post<number>(apis.create, erp_purchase_order_attachment);
}

export const updateErpPurchaseOrderAttachment = (erp_purchase_order_attachment: ErpPurchaseOrderAttachmentRequest): Promise<void> => {
  return api.post<void>(apis.update, erp_purchase_order_attachment);
}

export const deleteErpPurchaseOrderAttachment = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getErpPurchaseOrderAttachment = (id: number): Promise<ErpPurchaseOrderAttachmentResponse> => {
  return api.get<ErpPurchaseOrderAttachmentResponse>(`${apis.get}/${id}`);
}

export const listErpPurchaseOrderAttachment = (): Promise<Array<ErpPurchaseOrderAttachmentResponse>> => {
  return api.get<Array<ErpPurchaseOrderAttachmentResponse>>(apis.list);
}

export const pageErpPurchaseOrderAttachment = (condition: ErpPurchaseOrderAttachmentQueryCondition): Promise<PaginatedResponse<ErpPurchaseOrderAttachmentResponse>> => {
  return api.get<PaginatedResponse<ErpPurchaseOrderAttachmentResponse>>(apis.page, condition);
}

export const listErpPurchaseOrderAttachmentByPurchase = (id: number): Promise<Array<ErpPurchaseOrderAttachmentResponse>> => {
  return api.get<Array<ErpPurchaseOrderAttachmentResponse>>(`${apis.list_purchase}/${id}`);
}
